### PR TITLE
support authorize non root user to sso login

### PR DESCRIPTION
**Pull request by baurine**
What did:
- Support modify authorization at any time
- Support authorize to non-root user to login by sso

Demo:

image

image

image

image

### CODE REVIEW

File: pkg/apiserver/user/sso/service.go [hunk 1]
```diff
@@ -279,9 +283,6 @@ func (s *Service) discoverOIDC(issuer string) (*oidcWellKnownConfig, error) {
 		return nil, ErrDiscoverFailed.Wrap(err, "Failed to discover OIDC endpoints")
 	}
 	wellKnownConfig := resp.Result().(*oidcWellKnownConfig)
-	if wellKnownConfig.Issuer != issuer {
-		return nil, ErrDiscoverFailed.New("Issuer did not match in the OIDC provider, expect %s, got %s", issuer, wellKnownConfig.Issuer)
-	}
```
Comment (baurine) [inline pkg/apiserver/user/sso/service.go:284]: // TODO: add an explanation about why remove it.

File: pkg/apiserver/user/sso/service.go [hunk 2]
```diff
@@ -235,11 +234,16 @@ func (s *Service) createImpersonation(user string, password string) (*SSOImperso
...
 		EncryptedPass:         encryptedInHex,
 		LastImpersonateStatus: nil,
 	}
-	err = s.params.LocalStore.Clauses(clause.Insert{Modifier: "OR REPLACE"}).Create(&record).Error
+	// currently, we only support to authorize one sql user
+	err = s.revokeAllImpersonations()
+	if err != nil {
+		return nil, err
+	}
+	err = s.params.LocalStore.Create(&record).Error
```
Comment (breeswish) [inline pkg/apiserver/user/sso/service.go:246]: This is not concurrent safe.

**Review by breeswish — COMMENTED** [review]
The rest LGTM

File: pkg/apiserver/user/sso/service.go (hunk 1 above)
Comment (breeswish) [reply pkg/apiserver/user/sso/service.go:284]: Yeah, why do we remove it?

Comment (baurine) [reply pkg/apiserver/user/sso/service.go:284]: Reasons:

1. For auth0, the `issuer` filed has `/` suffix in the `http://[org].us.auth0.com/.well-known/openid-configuration` response:

   ```json
   {
     issuer: "https://[org].us.auth0.com/",
     authorization_endpoint: "https://[org].us.auth0.com/authorize",
     token_endpoint: "https://[org].us.auth0.com/oauth/token",
     ...
   }
   ```

   For okta, the `issuer` filed has no `/` suffix in the `http://[org].okta.com/.well-known/openid-configuration` response:

   ```json
   {
     issuer: "https://dev-9991837.okta.com",
     authorization_endpoint: "https://dev-9991837.okta.com/oauth2/v1/authorize",
     token_endpoint: "https://dev-9991837.okta.com/oauth2/v1/token",
     ...
   }
   ```

   With this check, the user needs to use different rules to fill the **OIDC Discovery URL**, it confuses the user.

2. Remove this check, it still works (I have tested), even the user only fill the **OIDC Discovery URL** with `[org].us.auth0.com` or `[org].okta.com`, no `http://` or `https://` prefix and `/` suffix.

    Btw, I think this check is a little contradicting, we use an issuer value, get the correct `openid-configuration` response, but the original value is not exactly the same as the value in the response, then we tell the user input the wrong value. If the original issuer value is wrong, it should not get the correct `openid-configuration` response. So I think the check is redundant, is it right?

File: pkg/apiserver/user/sso/service.go (hunk 2 above)
Comment (baurine) [reply pkg/apiserver/user/sso/service.go:246]: OK, let me have a look.

File: pkg/apiserver/user/sso/service.go (hunk 1 above)
Comment (breeswish) [reply pkg/apiserver/user/sso/service.go:284]: I think we should keep this check (and add more checks) according to OIDC specification:

https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfig

OpenID Provider Configuration Validation

If any of the validation procedures defined in this specification fail, any operations requiring the information that failed to correctly validate MUST be aborted and the information that failed to validate MUST NOT be used.

The issuer value returned MUST be identical to the Issuer URL that was directly used to retrieve the configuration information. This MUST also be identical to the iss Claim value in ID Tokens issued from this Issuer.

File: pkg/apiserver/user/sso/service.go (hunk 2 above)
Comment (baurine) [reply pkg/apiserver/user/sso/service.go:252]: Add mutex.

File: pkg/apiserver/user/sso/service.go (hunk 1 above)
Comment (baurine) [reply pkg/apiserver/user/sso/service.go:284]: Ok, let me have a look.

Comment (baurine) [reply pkg/apiserver/user/sso/service.go:284]: Revert and add more checks.

File: pkg/apiserver/user/sso/service.go [hunk 3]
```diff
@@ -254,16 +264,21 @@ func (s *Service) revokeAllImpersonations() error {
...
-	EndSessionURL string `json:"end_session_endpoint"`
+	Issuer                           string   `json:"issuer"`
+	AuthURL                          string   `json:"authorization_endpoint"`
+	TokenURL                         string   `json:"token_endpoint"`
+	UserInfoURL                      string   `json:"userinfo_endpoint"`
+	EndSessionURL                    string   `json:"end_session_endpoint"`
+	JWKSURI                          string   `json:"jwks_uri"`
+	ResponseTypesSupported           []string `json:"response_types_supported"`
+	SubjectTypesSupported            []string `json:"subject_types_supported"`
+	IDTokenSigningAlgValuesSupported []string `json:"id_token_signing_alg_values_supported"`
```
Comment (baurine) [inline pkg/apiserver/user/sso/service.go:275]: These fields are required according to the OIDC specification. https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata

**Review by breeswish — COMMENTED** [review]
Generally lgtm. @shhdgit PTAL, thanks

File: pkg/apiserver/user/sso/service.go (hunk 3 above)
Comment (breeswish) [reply pkg/apiserver/user/sso/service.go:275]: Could you help verify that newly added checks works for Okta? We have production users using Okta and when we adding more checks we need to make sure not breaking them.

File: pkg/apiserver/user/sso/service.go [hunk 4]
```diff
@@ -273,23 +288,35 @@ func (s *Service) discoverOIDC(issuer string) (*oidcWellKnownConfig, error) {
...
 		return nil, ErrDiscoverFailed.Wrap(err, "Failed to discover OIDC endpoints")
 	}
 	wellKnownConfig := resp.Result().(*oidcWellKnownConfig)
-	if wellKnownConfig.Issuer != issuer {
+	if strings.TrimSuffix(wellKnownConfig.Issuer, "/") != issuer {
 		return nil, ErrDiscoverFailed.New("Issuer did not match in the OIDC provider, expect %s, got %s", issuer, wellKnownConfig.Issuer)
 	}
 	if len(wellKnownConfig.TokenURL) == 0 {
-		return nil, ErrDiscoverFailed.New("TokenURL is not provided in the OIDC provider")
+		return nil, ErrDiscoverFailed.New("token_endpoint is not provided in the OIDC provider")
```
Comment (shhdgit) [inline pkg/apiserver/user/sso/service.go:301]: That's we validate many fields that must be implemented in the specification, `OpenID Provider Configuration Validation` only mentions `issuer url` and `iss Claim`?

And what is the `iss Claim value in ID Tokens issued`

File: pkg/apiserver/user/sso/service.go (hunk 3 above)
Comment (baurine) [reply pkg/apiserver/user/sso/service.go:275]: Yep, both checked.

File: pkg/apiserver/user/sso/service.go (hunk 4 above)
Comment (baurine) [reply pkg/apiserver/user/sso/service.go:301]: https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata mentions the required metadata.

**Review by breeswish — APPROVED** [review]

File: pkg/apiserver/user/sso/service.go (hunk 4 above)
Comment (shhdgit) [reply pkg/apiserver/user/sso/service.go:301]: https://github.com/coreos/go-oidc
https://connect2id.com/blog/how-to-validate-an-openid-connect-id-token

It looks like we should verify the issuer and issuer claims from id tokens in the exchange phase, not in the discovery phase.

Comment (shhdgit) [reply pkg/apiserver/user/sso/service.go:301]: Maybe we need to do the verification here:

https://github.com/pingcap/tidb-dashboard/blob/master/pkg/apiserver/user/sso/service.go?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L362

**Pull request merged**